publisher-sdk/src/main/java/com/criteo/publisher/logging/PublisherCodeSanitizer.kt

```diff
@@ -0,0 +1,220 @@
+        deepSanitize(it, visited)
+      }
+
+      with(ThrowableReflection) {
+        newException.realSuppressedExceptions = copySuppressed
+      }
+    }
+
+    val newStackTrace = original.stackTrace.fold(mutableListOf<StackTraceElement>()) { acc, stackTraceElement ->
+      if (stackTraceElement.isAllowed()) {
+        acc.add(stackTraceElement)
+      } else {
+        if (acc.isEmpty() || acc.last() != privateStackTraceElement) {
+          // Group the private stacks together
+          acc.add(privateStackTraceElement)
+        }
+      }
+      acc
+    }
+
+    newException.stackTrace = newStackTrace.toTypedArray()
+
+    return newException
+  }
+
+  /**
+   * Copy the type and the message of the given exception.
+   *
+   * [Throwable] is [Serializable], then it is possible to deeply copy it via serialization trick.
+   * But an exception can be really big in memory because of:
+   * 1. many cause or suppressed exceptions
+   * 2. embedded data (such as an Android [android.app.Application])
+   *
+   * 1. is solved by only doing a shallow copy of the throwable.
+   * 2. is solved if this method is only called on throwable thrown by the SDK. Then the SDK has an implicit
+   * responsibility to not put lot of data in its throwable.
+   */
+  private fun Throwable.shallowCopy(): Throwable {
+    var originalCause: Throwable?
+    var originalSuppressed: List<Throwable>?
+
+    // Save and remove sub exceptions
+    with(ThrowableReflection) {
+      originalCause = realCause
+      originalSuppressed = realSuppressedExceptions
+
+      realCause = null
+      realSuppressedExceptions = null
+    }
+
+    try {
+      // Clone
+      val bytes = ByteArrayOutputStream().use { baos ->
+        ObjectOutputStream(baos).use { oos ->
+          oos.writeObject(this)
+        }
+
+        baos.toByteArray()
+      }
+
+      val copy = ByteArrayInputStream(bytes).use { bais ->
+        ObjectInputStream(bais).use { ois ->
+          ois.readObject()
+        }
+      } as Throwable
+
+      // Add sub exceptions in copy
+      with(ThrowableReflection) {
+        // Throwable set "this" in its cause field to indicate that there is no cause
+        copy.realCause = if (originalCause == this@shallowCopy) {
+          copy
+        } else {
+          originalCause
+        }
+
+        copy.realSuppressedExceptions = originalSuppressed
+      }
+
+      return copy
+    } finally {
+      // Restore state
+      with(ThrowableReflection) {
+        realCause = originalCause
+        realSuppressedExceptions = originalSuppressed
+      }
+    }
+  }
+
+  private fun StackTraceElement.isAllowed(): Boolean {
+    return isSdk() || allowedPackagePrefixes.any { className.startsWith(it) }
+  }
+
+  private fun StackTraceElement.isSdk(): Boolean {
+    return className.startsWith(criteoPackagePrefix)
+  }
+
+  private fun Throwable.isThrownByPublisher(): Boolean {
+    stackTrace.forEach {
+      if (it.isSdk()) {
```

Comment:
`isSdk` means it is the SDK. `isAllowed` means that it is not the SDK but a core framework (coming from the list at top of file).

The logic of `isThrowByPublisher` is: if the first stack (from recent to old) that is not a core framework is not the SDK, then we consider it as a publisher exception (which can be thrown by the publisher itself or by a third party is using).

So the distinction between `isSdk` and `isAllowed` is important. But maybe I should rename the `allowedPackagePrefixes` to `allowedFrameworkPackagePrefixes`.

Reply:
I think renaming is not necessary. What would be useful is a high level concise description of what this code tries to achieve.

Reply:
> It appears that the most viable option (sanitized with reflection)takes around 3 seconds.
> Preparing and sending logs is done on a background thread and the user isn't waiting on it, so 3-5 secs isn't the end of the world as it is unlikely that the process will be killed within that time frame.

It's 3 milliseconds. And the preparation of the logs is done on the logger thread. So it might be the main thread. And honestly, I don't know if we can give the `Throwable` to the `SendingQueue`. The elements (with the Throwable) would be serialized (via JSON). But the benchmark shows that majority of time is in the serialization copy and in the `Log.printStackTrace`. So we would pay the serialization price in both cases.

> OTOH, I am worried because you benchmarked on a relatively recent and powerful device. What would be the benchmark results like on an older device? Btw, how many iterations of the benchmark did you run?

The important part is the relative time I guess. It might not scale linearly but I guess it might give a good approximation. If we want more precision, unfortunately, I don't have other devices. If you have, we can do a zoom so I can explain you how to run the benchmark (normally just a `./gradlew benchmark:connectedCheck` should do the job).

This JSON is generated by the benchmark: 
![image](https://user-images.githubusercontent.com/3816099/102112972-c2aadb80-3e38-11eb-9991-af59b3a6c520.png)

So that would be 1, but there is ~50 time runs  ¯\_(ツ)_/¯

> Naive question: instead of doing a shallow copy of the throwable, why don't we just build a string representing the stacktrace as we visit the throwable/cause chain? That would be less costly and in the end that will be sufficient for our purposes.

This can do the job but honestly, I would find it really hard to code and more harder to read. There is a lot of edge cases to take into account. Here is a few ones:
- printed stacktraces are cut: when you have the exception A caused by B, then the whole trace of A is printed, then you have something like "caused by B: message of B", then you have the trace of B until it joins the one of A (let's call the subtrace B'), and then you have `...`. So to determine if B belongs to the SDK or not, if the information is not in B', then you need to search in A from the separation stack of A and B. Same thing for suppressed. And then you need to handle this in a graph of stacktrace + recursion is stack
- throwable are a graph, not a tree. So you can have circular references, which is supported when printed and will show something like `[CIRCULAR REFERENCE exceptionA : exceptionB`]. Then good luck to determine to who is the exceptions

Even if we achieve to handle all of this (or detect & abandon in case of edge cases), then I'm not even sure that, with all the strings operations, it would be faster.

Reply:
Ah 3 millis, not sure why I saw 9-digits after the 3.  Ok, still, 3 millis on the main thread is a lot, it will cause skipped frames and janky user experiences.